File: src/tutorial/src/listener.cpp
#include <ros/ros.h>
#include <std_msgs/String.h>
using namespace ros;

class Listener
{
private:
  NodeHandle nh;
  Subscriber sub;

public:
  Listener()
  {
    sub = nh.subscribe("/chatter", 10, &Listener::callback, this);
  }

  void callback(const std_msgs::String::ConstPtr& str_msg)
  {
    ROS_INFO("heard %s",str_msg->data.c_str());
  }
};

int main(int argc, char **argv)
{
  init(argc, argv ,"listener");
  Listener listener;
  spin();
  return 0;
}

/*
void callback(const std_msgs::String::ConstPtr& str_msg){
    ROS_INFO("heard %s",str_msg->data.c_str());
}

int main(int argc, char **argv)
{
  ros::init(argc, argv, "listener");
  ros::NodeHandle nh;
  ros::Subscriber sub = nh.subscribe("chatter",10,callback);
  ros::spin();
  return 0;
}*/

File: src/tutorial/script/listener.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import rospy
from std_msgs.msg import String

class Listener:
    def __init__(self):
        self.sub = rospy.Subscriber("/chatter", String, self.callback)

    def callback(sef, data):
        rospy.loginfo(" I heard %s", data.data)

if __name__ == '__main__':
    rospy.init_node('listener', anonymous = True)
    Listener()
    rospy.spin()

'''
def callback(data):
    rospy.loginfo(rospy.get_caller_id() + " I heard %s", data.data)

def listener():
    rospy.init_node('listener', anonymous = True)
    rospy.Subscriber("chatter", String, callback)
    rospy.spin()

if __name__ == '__main__':
    listener()
'''
